Implement column title alignment for Tree

Barugon opened the pull request:
godotengine/godot-proposals#982

Implement methods `Tree::set_column_title_alignment` and `Tree::get_column_title_alignment`.

rcorre committed:
Disable local space for blender transforms.

Having local_space enabled when starting a transform changed the
behavior of VIEW space transforms. Now we disable local_space when
starting a blender transform (there was already logic to restore the
setting after the transform ends).

This also hides the gizmo while performing a blender transform,
otherwise the user will see it snap back and forth between the local and
global alignment. I think the transform looks cleaner with the gizmo
hidden anyways.

Fixes #59392.

snailrhymer committed:
Stop pasted child nodes being assigned an owner when previously unowned

Make copy and pasting match duplication's ownership transferral behavior by storing ownership information in the duplicated nodes on the node clipboard, then checking that information when setting owners for pasted nodes.

cooperra committed:
Have the Rename Node action use the targeted Node to determine the current undo/redo context

Formerly, we deduced context implicitly, but this failed and always used the global context instead of the context of the scene containing the Node.

This happened because the first argument to `add_do_method`, the SceneTreeEditor, is a descendant of Node and outside the current game scene's tree (it's part of the editor instead). This led the code in `EditorUndoRedoManager::get_history_id_for_object` to choose global context.

My solution is to explicitly use the renamed Node to deduce our context because it will always be in the current scene in this situation.

Fixes #67276

pkowal1982 committed:
Fix scrolling behaviour with low page value

pfertyk committed:
Fix RichTextLabel crash with out of bound exception

Fixes #68242.

EricEzaM committed:
Fix cancelling selection (pressing escape) while gizmo editing making uncommitted changes.
* Pressing escape while gizmo editing will discard the changes made during that edit 'session'

eskandrej committed:
Fix for deselecting item when select_mode == SELECT_ROW

YeldhamDev committed:
Defer invalidation on FileDialog nodes

resistor committed:
[GDScript] Perform update-and-assign operations in place when possible.

This turns two bytecode operations into one by using the assignment
destination directly as the output of the binary operator. This manifests
in operations like `+=`.

bruvzg committed:
Fix text server GDExtension build.

Scony committed:
Add initial navigation tests

dynics-chutchinson committed:
Replaced operating system alert dialog with a warning log message,
toggled by a project setting.
Fixes #73141

Commit:
Create a safe temporary file with is_backup_save_enabled

Sauermann committed:
Fix some ways to create inconsistent Viewport sizes

In the editor, it was possible to set the size of a `SubViewport` even
in cases where a parent `SubViewportContainer` had stretch enabled.

This PR disables editing a `SubViewport.size` while the parent disallows
it and it makes necessary adjustments during `NOTIFICATION_ENTER_TREE`.

bruvzg committed:
Add GDScript `to_wchar_buffer` and `get_string_from_wchar` functions.

bruvzg committed:
Automatically reparent editor message dialogs to avoid error spam.

MewPurPur committed:
Fix error spam when naming a func at the end of the script

bruvzg committed:
[macOS] Re-add support for the _sc_ inside app bundle. Update docs.

AThousandShips committed:
Fix TileSetEditor paiting texture_origin Vector2i

YuriSizov committed:
Reorganize context menu in FileSystem dock to put more used options higher

hakro committed:
Use physical shortcuts for freelook navigation in the editor

bruvzg committed:
[TextServer] Ensure ICU data is initialised only one and cleaned only at exit.

Calinou committed:
Use 8×8 default grid size for TextureRegion and 2D polygon editors

Power-of-two grid sizes are more suited to most game assets.

fabriceci committed:
Exposes the apply_floor_snap function to allow a snap to be made regardless of velocity.

bruvzg committed:
Improve layout direction/locale automatic selection.

SirQuartz committed:
Modify the default theme GraphNode close_h_offset

Modifies the close_h_offset to be 12 instead of 22. This better aligns the default position.

myaaaaaaaaa committed:
Fix data races in startup/teardown

aaronfranke committed:
Document GLTFNode and some of GLTFState

YeldhamDev committed:
Fix `GraphNode` resizing when its bottom border is too thin

kilojool committed:
Add missing virtual bind for control::get_tooltip

clayjohn committed:
Clean up OUTPUT_IS_SRGB redefinitions

TheSecondReal0 committed:
Draw Camera2D outlines as 2 point primitives instead of 4

RandomShaper committed:
Lift restriction that resource load thread requester has to be the initiator

RedMser committed:
Fix clipboard relying on focused window

jtnicholl committed:
Update documentation for Mesh and MeshInstance3D

Add missing descriptions for Mesh's virtual methods, and
MeshInstance3D's methods relating to blend shapes. Slight tweaks to
several MeshInstance3D method descriptions.

raulsntos committed:
C#: Always show "Create C# solution" option

Prevents ending up with an empty C# menu.
The option to create the C# solution no longer disappears, to avoid confusing users.
If an user tries to use it when a C# solution already exists they are warned that it will override their sln and csproj files.

KrisCatboy committed:
fix spotlight shadows in volumetric fog

Barugon commented:
Results using the attached project:

![Screenshot from 2023-02-25 13-21-05](https://user-images.githubusercontent.com/16503728/221381288-862bc888-7b80-48e7-81be-bb8dec703c58.png)
[test_project.zip](https://github.com/godotengine/godot/files/10832405/test_project.zip)

clayjohn committed:
Add warnings for unsupported features in mobile and gl_compatibility backends

the-brickster committed:
Commit Fix for GLES3 point size not working

bruvzg committed:
[TextServer] Add mutex for FreeType face creation/deletion operations.

bruvzg committed:
[Bitmap font] Assume outline size is `1` if it's not set, but channel for outline is defined.

samsface committed:
Clarify when Expression `get_error_text()` is updated

aaronfranke committed:
GLTF: Delete unused skeleton_to_node

BastiaanOlij committed:
Steering on VehicleBody is in radians while interface is in degrees. Improved documentation to clear this up.

lyuma committed:
basisu: Loop over the last mipmap lod

SaracenOne committed:
Stop toaster notification circle flickering when notifications are all hidden.

Abdul-AZ committed:
Fix glTF mesh importer not freeing nodes correctly on import

BastiaanOlij committed:
Move roughness limiter and sort into their own classes

sarice8 committed:
Document default values for transition type and ease type.

voidedWarranties committed:
Remove unused `ScriptLanguage` methods

bruvzg committed:
[TextEdit / IME] Check selection before deleting to avoid unnecessary error messages.

bruvzg committed:
Fix Windows to Windows SSH remote deploy. Fix Windows `execute` exit code.

Faless committed:
[MP] Fix orphan StringName on close.

Use a static StringName for the registered default interface name.

daBlesr committed:
TileSet editor was out of sync with TileMap and incorrectly overwrote old selected TileSet after an edit call with a null pointer.

KoBeWi committed:
Improve documentation of EditorExportPlugin

akien-mga committed:
Cleanup 3-to-4 renames, prevent common words replacements

Fixes #73505.
Fixes #73996.

KoBeWi committed:
Safeguard Makefile commands

akien-mga committed:
Add 3-to-4 renames for project settings in project.godot

In the ConfigFile format, the first subpath is the category and is not part
of the line that the regex would match.

Fixes #66125.

bruvzg committed:
[Font] Implement `get_char_from_glyph_index` function.

akien-mga committed:
Sync controller mappings DB with SDL2 community repo

Synced with gabomdq/SDL_GameControllerDB@436c7e3d54a57189ea0ab44d05f36b7cc7ea496c

bruvzg committed:
Fix GDExtensions library export when multiple architectures are set.

Commit:
Add feature check to require min Vulkan api version 1.0 on Android

groud committed:
Fix a crash in the GLB importer

snibk committed:
Redraw 2d viewport when guides are cleared

bruvzg committed:
[DisplayServer] Implement screen_get_pixel method for LinuxBSD/X11, macOS and Windows.

KoBeWi committed:
Add early return for changing TileMap properties

bruvzg commented on scene/gui/tree.cpp:
```diff
@@ -4045,7 +4045,24 @@ void Tree::_notification(int p_what) {
 
-					Vector2 text_pos = tbrect.position + Point2i(sb->get_offset().x + (tbrect.size.width - columns[i].text_buf->get_size().x) / 2, (tbrect.size.height - columns[i].text_buf->get_size().y) / 2);
+					Vector2 text_pos;
+					switch (columns[i].title_alignment) {
```
Left/Right should be swapped if `cache.rtl` is `true`.

akien-mga committed:
Minor typo and docs URL fixes

dalexeev committed:
Enable `RichTextLabel` context menu if selection is enabled

KoBeWi committed:
Check for type mismatch in PropertyTweener.from()

RedworkDE committed:
C#: Fix crash when errors occur before language initialization.

Barugon commented:
It seems to me that `HORIZONTAL_ALIGNMENT_LEFT` and `HORIZONTAL_ALIGNMENT_RIGHT` should be renamed to `HORIZONTAL_ALIGNMENT_BEGIN` and `HORIZONTAL_ALIGNMENT_END`, but that's a separate discussion.

iamjustkoi committed:
Fix small mistake in Window.xml

Fix small mistake in Window.xml

akien-mga committed:
i18n: Sync translations with Weblate

akien-mga committed:
Update CHANGELOG.md for 4.0 release

We couldn't hope to complete an in-depth curated changelog for such
a massive release, so we simply link to our in-depth release notes :)

Also adds back the 3.3, 3.4, and 3.5 curated changelogs. Those were
released from the 3.x branch so theoretically do not share a common
history with 4.0 (which started development after 3.2), but in practice
users will still perceive 4.0 as the natural upgrade to 3.5.

akien-mga committed:
Bump version to 4.0-stable \o/

4 years of development.
12,000 merged pull requests.
7,000 fixed issues.
1,500 individual contributors across engine and docs.

The Godot 4.0 release is by all metrics our biggest release so far.
No stone has been left unturned, all parts of the engine have been
modernized, refactored, overhauled, rewritten, redesigned.

Our work is far from done. Many areas still have significant known issues,
and will require focused work from all willing contributors to fix blocking
bugs, implement missing features, optimize for performance or compatibility,
and improve the user experience.

But Godot 4.0 marks the start of the new, modern Godot Engine, and a solid
foundation for us all to build upon. Future 4.x releases will come with a
much faster cadence, enabling us to iterate quickly on new features and
improvements to what we already provide.

To all of you who were involved in making Godot 4.0 what it is today, however
big or small your contributions were:

THANK YOU!

This was a massive undertaking, and you all participated in unique and
wonderful ways to build a free and open source game engine for everyone to
use and enjoy. You are breathtaking! <3

akien-mga committed:
Bump version to 4.1-dev

Can't stop, won't stop, they said, huh?

vonagam committed:
GDScript: Fix checking if a call is awaited in compiler

BastiaanOlij committed:
Use MSAA 2D texture in multipass tonemapper

BastiaanOlij committed:
Add XRServer.world_origin property

RandomShaper committed:
Fix crash in resource load

themancalledjakob committed:
improve documentation for Font and FontVariation

Co-authored-by: Hugo Locurcio <hugo.locurcio@hugo.pro>

timothyqiu committed:
Fix dock name lost translation after layout change

* After you click in the dock select panel
* After you load an editor layout

Chubercik committed:
Modify JSON.stringify so that it doesn't create unnecessary empty lines from empty arrays

MewPurPur committed:
Make GDScript number highlighting stricter

Maran23 committed:
Project Converter: Do not convert lines that start with a comment

Lines that start with # or // are ignored

timothyqiu committed:
Fix "Convert Full Project" button not translated

Also fixes a typo in the CHANGELOG.

timothyqiu committed:
Improve POT Generation dialog

* Avoid "property not found" warnings when adding a file for the first
  time.
* When no file is added, disable the Generate POT button instead of
  printing a warning.

bruvzg committed:
[Linux/X11] Check if required xkb functions exist before using it.

timothyqiu committed:
Hide internal settings from the classref

Default actions are no longer internal since we want to document them.
They are still hidden from the Project Setting dialog because we hid the
whole `input/` group manually.

Chaosus committed:
Fix randfn to prevent generating of nan values

rcorre committed:
Don't strip whitespace when converting 3to4.

Fixes #74204.

The style guide says

> Always use one space around operators and after commas

The 3to4 conversion tool currently strips space in certain scenarios.
I've updated it to add space whenever it is generating new code.
In any case where it substitutes existing code, it leaves it as-is.

For example, connect(a,b,c) becomes `connect(a, callable(b, c))`, because the converter is adding new commads/parens.

However, `xform(Vector3(a,b,c))` becomes `Transform * Vector3(a,b,c)` because it uses the user's original Vector3 string whole. If the user originally had `xform(Vector3(a, b, c))`, then it becomes `Transform * Vector3(a, b, c)`.

Ideally we'd always preserve original formatting, but this seems quite difficult, so I tried to preserve it where we can, but air on the side of following the style guide whenever we're transforming code.

AThousandShips committed:
Add keycode project conversion

mihe committed:
Add `pick_ray` parameter to extension binding of `intersect_ray`

MarcusElg committed:
Fix Camera2D position smoothing properties not being grouped

KoBeWi committed:
Discourage reusing Tweens

clayjohn committed:
Add a few more rendering timestamps

RedworkDE committed:
C#: Get singleton instances using the Core name

and-rad committed:
Fixed minor typo in turbulence documentation

timothyqiu committed:
Notify child controls when BackBufferCopy's rect changed

Calinou committed:
Document using `String.uri_encode()` with `OS.shell_open()`

hakro committed:
Get the unlit / unshaded extension when importing GLTF and set it when exporting

akien-mga committed:
FBX: Disable importer when canceling FBX2glTF setup

Pretty hacky solution but it's better than an infinite loop.

All this import setup needs to be redone, it's very difficult to properly
bail out from an invalid import without triggering reimport loops.

Also fix underline not visible at default editor scale in LinkButton.

Fixes #73319.

akien-mga committed:
Linux: Don't try to link system embree3 on unsupported archs

dalexeev committed:
Fix GDScript code style regarding colon

raulsntos committed:
C#: Encode GodotProjectDir as Base64 to prevent issues with special characters

clayjohn committed:
Add proper default texture filter and repeat modes for Canvas shaders in the OpenGL3 renderer

bruce965 committed:
Error on hint_normal_roughness_texture and hint_depth_texture outside of spatial shader

aleksrutins committed:
Validate that C# class names are valid identifiers

rcorre committed:
Correct superclass constructors in 3to4.

Fixes #70542.

The 3to4 conversion tool was not handling superclass constructors.
We should translate the godot3 syntax:

```gdscript
func _init(a,b,c).(a,b,c):
    pass

func _init(a,b,c):
    super(a,b,c)
```

Originally, the _init conversion was intended to remove `void` return types from _init functions, as this was disallowed due to #50589.
As that was resolved by #53366, I removed that part of the conversion logic. If a void return type is present on a constructor, the converter now leaves it.

Here's a sample diff from my own project:

```diff
@@ -103,10 +105,11 @@ class Real:
 class Text:
        extends Setting

-       var choices: PoolStringArray
-       var value: String setget set_value, get_value
+       var choices: PackedStringArray
+       var value: String : get = get_value, set = set_value

-       func _init(section: String, key: String, default: String, choice_list: Array).(section, key, default) -> void:
+       func _init(section: String, key: String, default: String, choice_list: Array) -> void:
+               super(section, key, default)
                choices = choice_list

        func normalize(val):
@@ -129,9 +132,10 @@ class Text:
 class Boolean:
        extends Setting

-       var value: bool setget set_value, get_value
+       var value: bool : get = get_value, set = set_value

-       func _init(section: String, key: String, default: bool).(section, key, default) -> void:
+       func _init(section: String, key: String, default: bool) -> void:
+               super(section, key, default)
                pass
```

JohanAR committed:
Fixed incorrect ERROR message when setting new PrimitiveMesh

If calling set_mesh with a PrimitiveMesh that has pending update, the
_mesh_changed function would be called twice. The first time before
set_base had been called, which could lead to an ERROR message about
trying to set an invalid surface override material.

JohanAR committed:
Fix instance uniforms breaking when setting a new mesh.

Fixes #58113

tlobig committed:
remove incorrect rename of get_used_cells_by_id

renaming get_used_cells_by_id to get_used_cells is not only unecessary, it introduces hard to debug issues

bitsawer committed:
Fix several clang-tidy bool literal conversion warnings

webair committed:
increased max touches to 32 for ios

raulsntos committed:
C#: Ignore explicit interface implementations

necrashter committed:
Use the new API for virtual keyboard height detection on Android, bugfix

SlugFiller committed:
Fix AABB calculation for meshes using Skeleton2D

rcorre committed:
Move tool declarations to top in 3to4.

In godot3, `tool` can follow keywords like `extends` and `class_name`
In godot4, `@tool` must be the first line in the file.

smix8 committed:
Fix NavigationServer free error print

Fixes error print for NavigationServer free when a RID can not be found.

rburing committed:
Fix concave/convex polygon shape documentation

Removes some outdated/incorrect/duplicated info, adds correct info.
Also removes no longer applicable "editor-only helper" statements.

akien-mga committed:
Windows: Workaround missing DWRITE_FONT_WEIGHT value in old MinGW

Fixes #74339.

isaaccp committed:
Clarify Thread documentation

The current documentation uses the term "running" for two different things.

In the description of get_id() it claims "it will return empty string if the thread is not running", but it actually will return the thread id until wait_to_finish is called.

In the description of is_alive() it claims "it will return true if the thread is running", but in this case it means "the provided function hasn't finished running yet".

Updated the functions documentation slighly to make this clear.

KoBeWi committed:
Fix EditorUndoRedoManager's handling of MERGE_ENDS

puchik committed:
Don't autocomplete numbers in GDScript

Cancel code autocompletion when a numerical value is entered. Avoids interference when setting numbers.

Commit:
Configure maven central snapshot versions for the Godot Android library

A snapshot version is a version that has not yet been released which allows us to deploy the same transient version incrementally, without requiring projects to upgrade the artifact version they're consuming. Those projects can use the same version to get an updated snapshot version.

bruvzg committed:
Fix IME position in the single window mode sub-windows.

akien-mga committed:
CI: Pin SCons to 4.4.0, the new 4.5.0 is broken

bruvzg committed:
[Windows] Update modifier key status during IME input.

bruvzg committed:
[Web] Fix some missing keycode mappings.

bruvzg committed:
[TextServer] Fix justification on punctuation characters.

RedworkDE committed:
C#: Fix building projects for MSBuild before 17.3

AThousandShips committed:
Fix type icons in `PropertySelector`

And adding a check to prevent future issues.

bruvzg committed:
[TextEdit] Do not draw virtual spaces (word break / justification points).

hakro committed:
Reload built in script to get its methods on signal connection

stmSi committed:
Fix `Find in Files` Search Results cannot open builtin script

Maran23 committed:
Convert OpenSimplexNoise to FastNoiseLite

- class name and octaves property

Maran23 committed:
Add conversion for Vector2 method tangent() -> orthogonal()

For C# also: Perpendicular() -> Orthogonal()

juliuse98 committed:
Fix blend-file import when using custom color management in blender

When using custom color management in blender the --version command will
output additional information infront of the "normal" output and this
prevented the import.

Fixes #74439.

clayjohn committed:
Set properties of ImageTexture3D when creating

rcorre committed:
Add some missing renames to 3to4 tool.

MultiplayerPeerExtension isn't an exact replacement for
NetworkedMultiplayerCustom, but at least it gets you moving in the right direction.

Engine.editor_hint couldn't be fixed by the renames map, because you have to add a `()` at the end.

clayjohn committed:
Avoid overflow when calculating ptr address for 3D textures in RenderingDevice texture update

ronyeh committed:
Rename mkv_list to mvk_list (MoltenVK path list).

SlugFiller committed:
Fixes a canvas item set to clip children being drawn as black if no children are visible

Rindbee committed:
Fix broken shortcut key input

nongvantinh committed:
Fixes  gutter set width results in receiving only half of the desired size

YuriSizov committed:
Add missing handler for removing font sizes in Themes

YuriSizov committed:
Generate empty textures for theme icons if the SVG module is disabled

RedworkDE committed:
Fix buffer overrun in CPUParticles3D

AThousandShips committed:
[TextServer] Make sure Freetype is enabled for ot-svg

YuriSizov committed:
Improve logic related to editing audio buses (and prevent crashes)

clayjohn committed:
Propogate errors when creating an OpenGL context fails in X11

YuriSizov committed:
Update property keying state without a full Inspector rebuild

YuriSizov committed:
Prevent crashing on startup if project has scripted theme types

Also avoid order of operation conflicts by moving C# binding generation hook to main.cpp

clayjohn committed:
Avoid copying CanvasTexture when updating proxy

timothyqiu committed:
Fix crash when dumping extension API in a non-writable directory

Haydoggo committed:
Fix crash when revealing file in floating FileSystem Dock
When selecting "Show in FileSystem" from the context menu of a resource
in the inspector, the engine would crash if the FileSystem dock was
floating because it was trying to focus the FileSystem tab, but floating
docks don't use Tab Containers. This commit makes the FileSystem dock's
window grab focus instead if it's floating.

bruvzg committed:
[Linux/X11] Add some missing keycodes/scancodes.

timothyqiu committed:
Fix coloring of the renderer options button

rcorre committed:
Don't unset local_coords on starting transform.

This fixes a bug introduced by 806425621cfa3daba3ba9f6b020726490a3152a4, where dragging the gizmo no longer respected local transforms.

I'm not sure why I called set_local_coords_enabled(false) in _compute_edit. Removing this line seems to fix gizmo-dragging local transforms, without breaking anything else.

I also noticed that confirming a transform leaves the gizmo axis lines on the screen. This is fixed by calling update_transform_gizmo after clearing the edit mode/instant flags, so update_transform_gizmo knows not to render any axes.

AThousandShips committed:
Fix Variant hashing for floats

Incorrectly hashed floats as single precision

brunosxs committed:
Fix for AABB methods with wrong description

vnen committed:
Remove outdated note about const in Dictionary and Array docs

YuriSizov committed:
Prevent cache corruption when saving resources in the editor

vonagam committed:
GDScript: Change parser representation of class extends

vonagam committed:
GDScript: Fix missing warning for shadowing of built-in types

Calinou committed:
Tweak command syntax in Vulkan renderer failure message

This quotes the executable name so that copying it always works
(even if the path contains spaces).

The command is also indented from the rest of the text and is
no longer single-quoted, as that can prevent the command from
running if the line is copied in its entirety (with the quotes).

Maran23 committed:
Add conversion for common Theme Overrides

titus125 committed:
Fixed typos in the Transform2D and Transform3D class reference

KoBeWi committed:
Select the newly duplicated file

Haydoggo committed:
Fix errors when closing floating docks

dock->get_index() on line 4463 was not behaving as expected
due to dock having an internal sibling, so now we just get the
index excluding internal nodes.

line 4742 would throw an error if you made multiple docks
floating then redocked the end docks first, but no longer

timothyqiu committed:
Add missing TTRs in tiles editor and array inspector

timothyqiu committed:
Fix various typos in the classref

omar-polo committed:
change some sys.exit() to be fatal

- when not all freetype deps are builtin (or provided externally)
- when attempting to build on an unsupported CPU architecture

akien-mga committed:
Vulkan: Fix VMA build with GCC 13

Fixes #74647.

YuriSizov committed:
Prevent passing events from CodeEdit to TextEdit when code completion is active

akien-mga committed:
SCons: Cleanup pulseaudio defines for Linux

No need to define _REENTRANT manually when using the system lib,
it's part of the pkgconfig cflags.
And we were then defining PULSEAUDIO_ENABLED twice.

YuriSizov committed:
Prevent infinite loops when printing orphan nodes

RedworkDE committed:
Add dedicated setter for `ImageTextureLayered::_images` to fix `create_from_images` being hidden from the C# bindings

and-rad committed:
Shader editor trims trailing whitespace if set in editor settings

mihe committed:
Fix extension bindings for motion collision/result structs

lewiji committed:
Add "filesRoot" path to Android provider paths xml

AThousandShips committed:
Document `editor/naming/scene_name_casing` setting

Moved definitions of editor related project settings to `editor/register_editor_types.cpp` to make documentation work.

clayjohn committed:
Delete unused compression formats from .imoprt files when exporting

vonagam committed:
GDScript: Fix autocomplete inside a block with a type test condition

clayjohn committed:
Add parentheses around arguments when converting xform

bruvzg committed:
[TextServer] Add invalid font scaling check, restrict Linux/BSD system fonts lookup to TrueType/CFF only.

mihe committed:
Add `recovery_as_collision` to extension binding of `_body_test_motion`

BastiaanOlij committed:
Merge duplicate rd_texture functions

yedpodtrzitko committed:
clear filter input in Project Settings when opening Layers

AThousandShips committed:
Exposing more project settings for documentation

brettchalupa committed:
Fix CharacterBody2D get_slide_collision docs

The indentation was off and the code usage was incorrect for Godot 4.0 GDScript.

justchen1369 committed:
Fix exported type for Menubar start_index

YuriSizov committed:
Make sure Script Debugger is updated with the editor theme

KoBeWi committed:
Remember directory when installing templates file

clayjohn committed:
Use linear filtering without mipmaps for ProceduralSkyMaterial and PhysicalSkyMaterial

AThousandShips committed:
Fix type check for max/min

Malcolmnixon committed:
Fix null-pointer dereference on Android systems when using gl_compatibility renderer.

luzader committed:
OpenXR: Enable access to the Valve Index grip force sensing

KoBeWi committed:
Fix MultiNodeEdit not cleared after deleting nodes

smix8 committed:
Allow negative NavigationAgent2D path debug line_width for thin lines

Allows the line_width for NavigationAgent2D path debug to go negativ for thin line rendering.

JohnVeness committed:
Corrected capitalisation of macOS

In documentation and comments, ignoring thirdparty code

dsnopek committed:
Documents the "arc-based" Quaternion constructor

pcamp committed:
Fix typo in Window class.

"event as" -> "events will be"

BastiaanOlij committed:
Add a get_system_info method to XRInterface

dalexeev committed:
Fix `InputEventConfigurationDialog` modifies original event

lawnjelly committed:
Improve Vector2 / 3 / 4 normalized() classref.

Mention that the results will be unreliable with initial vector near zero.

rburing committed:
Revert attempted fix of trimesh CCD

fahadshihab committed:
Added drag and drop support for shaderinclude

Dragging and dropping now works for shader include file.

smix8 committed:
Fix GridMap free navigation RID error spam

Fixes GridMap free navigation RID error spam.

smix8 committed:
Fix GridMap signal "cell_size_changed" disconnect error

Fixes GridMap signal "cell_size_changed" disconnect error.

BastiaanOlij committed:
XR: When an sRGB target is used, check hardware sRGB conversion

smix8 committed:
Fix agents with disabled avoidance getting added to avoidance simulation

Fixes that agents with disabled avoidance were getting added to avoidance simulation.

Spartan322 committed:
Remove methods in header removed in c7eb814

c7eb814 missed these methods in the RichTextLabel's header
Remove set_override_selected_font_color from rich_text_label.h
Remove is_override_selected_font_color from rich_text_label.h
c7eb814 missed these methods in the TextEdit's header
Remove set_override_selected_font_color from text_edit.h
Remove is_override_selected_font_color from text_edit.h

YuriSizov committed:
Make the request to redraw when clearing guides a part of UndoRedo

AThousandShips committed:
Fix `Array.slice()` rounding for `abs(step) != 1`

paddy-exe committed:
Fix ndc calculation for LinearSceneDepth VS node in GLES3

bruvzg committed:
[macOS] Remove unnecessary debug prints.

Calinou committed:
Add "Bbcode" -> "BBCode" editor capitalization

This property name is used in RichTextLabel.

Rindbee committed:
Property list changes are only notified when it did change in Curve

`Object::notify_property_list_changed()` should only be called when
the structure of the property has actually changed.

The structure of the property tree has not changed if the size of the
array has not changed, and the only possibility is that the element
value has changed.

akien-mga committed:
i18n: Sync translations with Weblate

(cherry picked from commit d23922ffebe48f29126c003411495737d07e5a9f)

bruvzg committed:
Remove (or make verbose only) various debug prints.

dalexeev committed:
GDScript: Fix false positive `REDUNDANT_AWAIT` warning

clayjohn committed:
Disallow creating a project in the Home or Documents folder

Also don't prompt users to delete the project folder for now.

This is a temporary fix to ensure that users do not delete their entire home folders by mistake

timothyqiu committed:
Fix typos and inconsistencies in classref

timothyqiu committed:
Fix error when opening Inspector's dots menu

When the inspector is empty, opening the "extra resource options" menu
produces a error:

> Condition "current_res.is_null()" is true.

Opening the mneu with no current resource is valid use case, so
`ERR_FAIL_COND` should not be used.

smix8 committed:
Fix NavigationAgent3D debug path duplicated points

Fixes duplicated points in NavigationAgent3D debug path.

kleonc committed:
TileMapEditor Fix preview rendering and transform calculations

KoBeWi committed:
Fix auto-translations in editor

bruvzg committed:
[RTL] Add option to customize list bullet, use U+2022 by default.

mihe committed:
Change documentation of `Shape3D.margin` to reflect current situation

mihe committed:
Open up bounds of `Shape3D.margin` to support other physics engines

MisoMosiSpy committed:
Fix typo in the type-safe examples

Fix type in the type-safe examples of snapped, lerp and sign function.

bruvzg committed:
Fix Unix temp file creations when using is_backup_save_enabled.

adamscott committed:
Set VehiculeWheel3D `suspension_travel` default value to a resonable one

Commit:
Make vulkan level 1 an optional requirement

Calinou committed:
Improve Color8 documentation

This documents in which situations Color8 is most suited and its
precision limitations.

dsnopek committed:
Add missing documentation for MultiplayerPeerExtension

KoBeWi committed:
Fix get_drag_data not overridable in some Controls

Rindbee committed:
Add instructions for configuring CheckBox as a radio button

Commit:
Fix directory access when the running app has the `All files access` permission

Pennycook committed:
Add get_loops_left() function to Tween

Implements godotengine/godot-proposals#5141.

Adds a new get_loops_left() function to Tween, allowing developers to
reason about how many times a tweening sequence will repeat and whether
to expect finished or loop_finished as the next signal.

Co-authored-by: Tomek <kobewi4e@gmail.com>

bruvzg committed:
[Windows] Remove meta_mem update on keyup/keydown.

Rindbee committed:
Fix some wrong descriptions in the docs

ator-dev committed:
Fix "Download Project Source" for Web Editor

- Correctly pass safe project name and safe datetime to `vformat`

AThousandShips committed:
Make `Gradient` resort points on `reverse`

nikitalita committed:
Fix ResourceImporterLayeredTexture::import() high_quality type

AThousandShips committed:
Improve documentation for Area monitor callbacks in `PhysicsServer3D`

AThousandShips committed:
Fix `AnimatedSprite2D` autoplay warning

Changing of autoplay when changing `SpriteFrames` is not done by the
user and warning is unhelpful

rburing committed:
Fix collide_shape return type

Fix PhysicsDirectSpaceState3D::_collide_shape return type.
Also PhysicsDirectSpaceState2D::_collide_shape.

RedworkDE committed:
C#: Do not print errors about missing references to intentionally ignored members

AThousandShips committed:
Fix "hides member" warnings in `csharp_script.cpp`

Barugon commented:
Any chance of getting this in v4.1?

Barugon commented:
Windows failure doesn't look like it's due to my changes.

KoBeWi commented:
I think you need to rebase the PR to fix it.

KoBeWi commented on doc/classes/Tree.xml:
```diff
@@ -288,6 +295,14 @@
+			<param index="0" name="column" type="int" />
+			<param index="1" name="title_alignment" type="int" enum="HorizontalAlignment" />
+			<description>
+				Sets column title alignment.
```
Seems like column titles don't support newlines, so `HORIZONTAL_ALIGNMENT_FILL` won't have effect here. Might be worth mentioning that maybe.

Barugon replied:
Should setting `HORIZONTAL_ALIGNMENT_FILL` produce an error message?

KoBeWi replied:
Maybe it could print a warning 🤔

Barugon replied:
Perhaps something like this?
```C++
	if (p_alignment == HORIZONTAL_ALIGNMENT_FILL) {
		WARN_PRINT("HORIZONTAL_ALIGNMENT_FILL is not supported for column titles");
		return;
	}
```

KoBeWi replied:
I think the warning should be in `set_text_alignment()` if anything.

KoBeWi reviewed (approved):
Well aside the note about fill alignment, seems fine. Needs squashing tho.

Barugon commented:
> Needs squashing tho.

I don't know how to accomplish that.

Barugon commented:
Well, that didn't work so good.

Barugon commented:
I'm going to close this PR ad redo it.

KoBeWi commented:
This is why you should never use the GitHub buttons to rebase xd
